### PR TITLE
feat: Add Scopes to KeycloakClient Authorization spec (#41)

### LINKED ISSUE #41: Add Scopes to Keycloak Client Authorizations
**Is your feature request related to a problem? Please describe.**
When using the new Authorization field for KeycloakClient (Thanks for adding that feature), if the permission is scoped based, then the scope needs to exist in the Authorization Scopes for the Client.

**Describe the solution you'd like**
Add a Scopes field to the KeycloakClient Authorization spec

**Describe alternatives you've considered**
Manually adding the authorization scope e.g "token-exchange" allows the KeycloakClient to reconcile properly

**Additional context**
I am working on a branch for this now, but I'm currently stuck at the mock pieces because `make mocks` is hitting a nil pointer. If you could possibly point into the right direction of which piece I could have missed, I would love to get this PR in for your team.

### PR DESCRIPTION
**Pull request by dougkirkley**
# Pull Request Template

## Description
When using the new Authorization field for KeycloakClient (Thanks for adding that feature), if the permission is scoped based, then the scope needs to exist in the Authorization Scopes for the Client.

Fixes #41

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Pull Request contain one commit. I squash my commits

### CODE REVIEW

**Comment by dougkirkley** [comment]
Currently trying to get all my mocks added and running into some issues. No idea where the first code block is coming from, because that function signature is not in the client interface 

```
CreateScope(context.backgroundCtx,string,string,string,gocloak.ScopeRepresentation)
        		0: context.backgroundCtx{emptyCtx:context.emptyCtx{}}
        		1: "token"
        		2: "rl"
        		3: "cl"
        		4: gocloak.ScopeRepresentation{DisplayName:(*string)(nil), IconURI:(*string)(nil), ID:(*string)(nil), Name:(*string)(0x1400004ce20), Policies:(*[]gocloak.PolicyRepresentation)(nil), Resources:(*[]gocloak.ResourceRepresentation)(nil)}
        
        The closest call I have is: 
        
        CreateScope(string,string,string,string,string)
        		0: "mock.Anything"
        		1: "token"
        		2: "rl"
        		3: "clid1"
        		4: "token-exchange"
```

**Comment by zmotso** [comment]
Ensure you create mocks by `mocks.NewMockClient(t)`.

Test example:
```go
func TestProcessScope_Serve(t *testing.T) {
	m := mocks.NewMockClient(t)
	m.On("GetClientID", "test", "realm").Return("clientID", nil)
	m.On("GetScopes", mock.Anything, "realm", "clientID").Return(nil, nil)
	m.On("CreateScope", mock.Anything, "realm", "clientID", "scope").Return(nil, nil)

	h := NewProcessScope(m)
	kc := &keycloakApi.KeycloakClient{
		Spec: keycloakApi.KeycloakClientSpec{
			ClientId: "test",
			Authorization: &keycloakApi.Authorization{
				Scopes: []string{"scope"},
			},
		},
	}

	err := h.Serve(context.Background(), kc, "realm")
	require.NoError(t, err)
}
```

**Comment by dougkirkley** [comment]
Okay, I was just copying the test from `TestGoCloakAdapter_AddDefaultScopeToClient` originally.

**Comment by dougkirkley** [comment]
@zmotso I figured out the testing, thanks for pointing me to the actual chain testing, I was trying to add it to the adapter tests.

**Comment by dougkirkley** [comment]
@zmotso This is ready for review now

File: controllers/keycloakclient/chain/process_scope_test.go [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+				},
+			},
+			keycloakApiClient: func(t *testing.T) keycloak.Client {
+				client := mocks.NewMockClient(t)
+				client.On("GetClientID", "client", "master").
+					Return("clientID", nil).Once()
+				client.On("GetScopes", mock.Anything, "master", "clientID").
+					Return(map[string]gocloak.ScopeRepresentation{
+						"scope-auth": {
+							Name: gocloak.StringP("scope-auth1"),
```
Comment (zmotso) [inline controllers/keycloakclient/chain/process_scope_test.go:52]: Add `ID` property.
Without ID you have panic here
`if err := h.keycloakApiClient.DeleteScope(ctx, realmName, clientID, *existingScopes[name].ID)`

File: controllers/keycloakclient/chain/process_scope_test.go [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+				},
+			},
+			keycloakApiClient: func(t *testing.T) keycloak.Client {
+				client := mocks.NewMockClient(t)
+				client.On("GetClientID", "client", "master").
+					Return("clientID", nil).Once()
+				client.On("GetScopes", mock.Anything, "master", "clientID").
+					Return(map[string]gocloak.ScopeRepresentation{
+						"scope-auth": {
+							Name: gocloak.StringP("token-exchange"),
```
Comment (zmotso) [inline controllers/keycloakclient/chain/process_scope_test.go:78]: The same here.

File: pkg/client/keycloak/adapter/gocloak_adapter_client.go [hunk 3]
```diff
@@ -119,6 +119,28 @@ func (a GoCloakAdapter) GetResources(ctx context.Context, realm, idOfClient stri
 	return resources, nil
 }
 
+// CreateScope creates a client authorization permission.
+// nolint:gocritic // gocloak is a third party library, we can't change the function signature
+func (a GoCloakAdapter) CreateScope(ctx context.Context, realm, idOfClient string, scope string) (*gocloak.ScopeRepresentation, error) {
```
Comment (zmotso) [inline pkg/client/keycloak/adapter/gocloak_adapter_client.go:124]: Run `make lint` and fix couple of errors.

**Review by zmotso — COMMENTED** [review]
@dougkirkley  Thank you for your PR. The code looks good overall. Please fix a couple of comments.

Also, please change the commit message to
feat: Add Scopes to KeycloakClient Authorization spec (#41)

**Review by zmotso — APPROVED** [review]

**Comment by zmotso** [comment]
@SergK Please check this PR. We can merge it.

**Comment by dougkirkley** [comment]
Not sure why the commit message check is failing, it _looks_ right to me

**Comment by SergK** [comment]
@dougkirkley yep issue is on our side, I will align PR title to our pipelines, and later we will fix CI in a proper way

**Comment by SergK** [comment]
@dougkirkley Thank you for contribution, we'll fix CI on our side

**Pull request merged**